packages/hardhat-viem/src/internal/chains.ts

```diff
@@ -0,0 +1,94 @@
+import type { EthereumProvider } from "hardhat/types";
+import type { Chain } from "viem";
+import type { TestClientMode } from "viem/src/clients/createTestClient";
+
+import memoize from "lodash.memoize";
+
+import {
+  UnknownDevelopmentNetworkError,
+  NetworkNotFoundError,
+  MultipleMatchingNetworksError,
+} from "./errors";
+
+export async function getChain(provider: EthereumProvider): Promise<Chain> {
+  const chains: Record<string, Chain> = require("viem/chains");
+  const chainId = await getChainId(provider);
+
+  if (isDevelopmentNetwork(chainId)) {
+    if (await isHardhatNetwork(provider)) {
+      return chains.hardhat;
+    } else if (await isFoundryNetwork(provider)) {
+      return chains.foundry;
+    } else {
+      throw new UnknownDevelopmentNetworkError();
+    }
+  }
+
+  const matchingChains = Object.values(chains).filter(
+    ({ id }) => id === chainId
+  );
+
+  if (matchingChains.length === 0) {
+    throw new NetworkNotFoundError(chainId);
```

Comment:
I feel like here we should repeat the `isHardhatNetwork`/`isFoundryNetwork` checks, to handle the case where someone is using a development network with a different chain id. This is not an uncommon practice.